### ISSUE TITLE
rspamd marks outgoing mail as spam and breaks DKIM signature

### ISSUE THREAD

**Issue by DrMurx** [hardware/mailserver#311]
#### Classification

* [x] Minor bug

#### Reproducibility

* [X] Always

#### Docker information

Image is `hardware/mailserver:1.1-stable` (ID `b12411bf5132`)

#### Description

rspamd seems to process outgoing mails too; in my case, the typical test email to send to others to confirm that things are working (with subject `test` and body `test`) is considered as spam by the rspamd classificator.

Therefore, in my outgoing emails, the subject of such emails is rewritten as `*** SPAM ***: test`. No other spam headers are added, though.

Since the emails are DKIM signed, the alteration of the subject renders the DKIM signature invalid for the receivers side. Depending on the frequency of those events, it might lower the reputation of the sender's IP on the receiver's side.

#### Steps to reproduce

1. Train rspamd's bayes filter with some reproducible spam 
2. Send an email like that to someone

#### Expected results

Subject left as-is & DKIM signature valid.

#### Actual results

The subject rewritten as `*** SPAM ***` and the DKIM signature broken.

#### Debugging information

Nothing about the issue can be seen in the logs.

The rspamd history tab in the UI confirms that subject has been rewritten on those outgoing emails.

**Comment by hardware** [comment]
I don't know why rspamd rewrite subject for outgoing mail and I did not find a simple way to disable this behavior, the only recommended way seems to apply a specific action for all authenticated users like this :

```ini
# https://rspamd.com/doc/configuration/settings.html
# https://rspamd.com/doc/configuration/metrics.html

authenticated {
  priority = high;
  authenticated = yes;
  apply {
    actions {
      "rewrite subject" = 100.0;
      # I tried this too :
      # rewrite_subject = 100.0;
      # subject = "";
    }
  }
}
```

But it does not work. I even tried to disable **all checks** for authenticated users :

```ini
authenticated {
  priority = high;
  authenticated = yes;
  want_spam = yes;
}
```

Unsuccessfully...

@vstakhov : Is it possible to disable subject rewriting for outbound emails ?

**Comment by hardware** [comment]
@jaen Thank you very much for this information.

**Comment by vstakhov** [comment]
It is not related to Rspamd development. Rspamd does not rewrite any subjects.

**Comment by hardware** [comment]
> When I put it in modules.d/settings.conf instead, it worked.

@jaen it doesn't work for me, I tried with `modules.d/settings.conf` without success.

@DrMurx If rspamd does not write any subjects as its creator claims, I have absolutely no idea what component can cause this bahavior.

**Comment by DrMurx** [comment]
@vstakhov

> Rspamd does not rewrite any subjects.

I'm confused - rspamd's history shows action "rewrite subject", rspamd's throughput graph shows "Subject rewritten". Yet you say rspamd doesn't rewrite any subjects?

**Comment by jaen** [comment]
@hardware
Well, I don't have this _exact_ config as you do – I have a different `apply`. But otherwise this settings override works and I just kind of assumed whatever you put in `apply` was correct. Sorry for confusion.

What I have exactly is this file at `./files/rspamd/local.d/settings.conf` relative to my compose:
```ini
settings {
  authenticated {
    priority = high;
    authenticated = yes;
    apply {
      groups_disabled = ["fuzzy"];
    }
  }
}
```
And I have that file mounted like so in the compose:
```
./files/rspamd/local.d/settings.conf:/etc/rspamd/modules.d/settings.conf
```
And this stops rewriting for me (using `1.1-stable1 tag). If it doesn't for you then that's weird.

**Comment by hardware** [comment]
@jaen It works because you disable all symbols in fuzzy group :

```ini
# scores.d/fuzzy_group.conf

symbols = {
    "FUZZY_UNKNOWN" {
        weight = 5.0;
        description = "Generic fuzzy hash match, bl.rspamd.com";
    }
    "FUZZY_DENIED" {
        weight = 12.0;
        description = "Denied fuzzy hash, bl.rspamd.com";
    }
    "FUZZY_PROB" {
        weight = 5.0;
        description = "Probable fuzzy hash, bl.rspamd.com";
    }
    "FUZZY_WHITE" {
        weight = -2.1;
        description = "Whitelisted fuzzy hash, bl.rspamd.com";
    }
}
```

And in @DrMurx's case, `FUZZY_DENIED` is triggered 

```
FUZZY_DENIED
Denied fuzzy hash, bl.rspamd.com
Score +12
```

but I do not think that's the best way to disable this behavior, I want to disable subject rewriting for outgoing mails, not all fuzzy mechanism.

@DrMurx He is right, **by default** rspamd rewrite any subject. Rspamd does it because I ask him to do it when score is above **8** :

```ini
# local.d/actions.conf 

reject = 20;
rewrite_subject = 8;
add_header = 6;
greylist = 4;
subject = "*** SPAM *** %s";
```

The weight of `FUZZY_DENIED` is 12, which is explain why the subjet is rewritten.

But that doest not explain why this :+1:  

```
authenticated {
  priority = high;
  authenticated = yes;
  apply {
    actions {
      "rewrite subject" = 100.0;
      # or
      rewrite_subject = 100.0;
      # or
      subject = "";
    }
  }
}
```

does not work.

**Comment by jaen** [comment]
Ah yeah, it's a bit of a nuclear option to be sure. It's certainly better for them to be kept analysed, just not rewritten. Hopefully this can be done somehow, because documentation seems to imply what you did should've worked.

**Comment by hardware** [comment]
Same behavior with Rspamd 1.8.1

@moisseev @fatalbanana Is it possible to disable subject rewriting for outbound emails ?

I already tried this settings without success :

```ini
authenticated {
  priority = high;
  authenticated = yes;
  apply {
    actions {
      "rewrite subject" = 100.0;
      # or
      rewrite_subject = 100.0;
      # or
      subject = "%s";
    }
  }
}
```

**Comment by fatalbanana** [comment]
Way to do it could be to remove `rewrite_subject` from `local.d/actions.conf` and add it back via settings when it's desirable (ie. not outbound).

**Comment by vstakhov** [comment]
> @vstakhov
> 
> > Rspamd does not rewrite any subjects.
> 
> I'm confused - rspamd's history shows action "rewrite subject", rspamd's throughput graph shows "Subject rewritten". Yet you say rspamd doesn't rewrite any subjects?

Exactly. Rspamd just tells MTA to perform subject rewriting. In theory, you can scan a message twice: once for spam checks (and doing this subject rewriting) and one for dkim signing where Rspamd would see an altered message. However, I personally don't understand who needs rewrite subject action and why. And this is a clear sample why it is bad.

**Comment by dellarte** [comment]
Hi @hardware @vstakhov ,

I also removed the local.d/actions and my outgoing mail are still sent with the subject rewrote * * * SPAM * * *

**Comment by moisseev** [comment]
You have probably left `rewrite` action in the dynamic configuration ("$DBDIR/rspamd_dynamic").

**Comment by dellarte** [comment]
Hi @moisseev ,

We're running rspamd 2.1.

The /etc/rspamd/actions.conf is left unmodified : 
``` 
actions {
    reject = 15; # Reject when reaching this score
    add_header = 6; # Add header when reaching this score
    greylist = 4; # Apply greylisting when reaching this score (will emit `soft reject action`)

    #unknown_weight = 1.0; # Enable if need to set score for all symbols implicitly
    # Each new symbol is added multiplied by gf^N, where N is the number of spammy symbols
    #grow_factor = 1.1;
    # Set rewrite subject to this value (%s is replaced by the original subject)
    #subject = "***SPAM*** %s"

    .include(try=true; priority=1; duplicate=merge) "$LOCAL_CONFDIR/local.d/actions.conf"
    .include(try=true; priority=10) "$LOCAL_CONFDIR/override.d/actions.conf"
}
```


In $DBDIR/rspamd_dynamic, we just set the score for rewrite subject on ingoing mails : 

```
[
    {
        "metric": "default",
        "actions": [
            {
                "name": "reject",
                "value": 15.0
            },
            {
# if we comment out "rewrite subject", rspamd doesn't rewrite either ingoing and outgoing mails
                "name": "rewrite subject",
                "value": 12.0
            },
            {
                "name": "add header",
                "value": 8.0
            },
            {
                "name": "greylist",
                "value": 5.0
            }
        ]
   }
]
```

**Comment by moisseev** [comment]
Why do you think you set it only to incoming? Actually, these thresholds are applied to every scanned e-mail. Rspamd has no idea which mail is outbound. If you really need to do it, you should split up mail flows, for instance [using `settings-id` and a separate MTA for outbound mail](https://rspamd.com/doc/faq.html#how-to-distinguish-inbound-and-outbound-traffic-for-rspamd-instance).
But if you want to keep it simple, just disable subject rewriting. I personally don't see any sense in messing with subject.
